### PR TITLE
Feature/54

### PR DESCRIPTION
**Pull request by StefanDywersant**
#54 Create UI for assigning/removing roles to/from given user. Created domain object which determines whether a role can be assigned/removed to/from some user based on assigner roles. This is also reflected in UI - you don't see roles which you can't assign and also not all roles can be removed.

- you need to rebuild opuscapita/user-base container because there were some changes
- the webpage is available here http://localhost:8080/user
- logged userId is configured in www/index.js
- edited userId is configured in www/App.react.js

### CODE REVIEW

File: src/client/components/UserRoleEditor/UserRoleTable.react.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import React, { Component, Children } from 'react';
+
+const UserRoleTable = ({ children }, context) => (
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleTable.react.js]: Please do not create components for structures that are too simple to gain any enhancements and are used only once in the whole project. Just put the HTML where it is actually used without creating additional structure complexity.

File: src/client/components/UserRoleEditor/UserRoleTableItem.react.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleTableItem.react.js]: Please do not create components for structures that are too simple to gain any enhancements and are used only once in the whole project. Just put the HTML where it is actually used without creating additional structure complexity.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	};
+
+	/**
+	 * Add role form submit handler
+	 * @param {object} event Form submit event object
+	 */
+	onSubmit = (event) => {
+		event.preventDefault();
+
+		if (!this.state.selectedRoleId) {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: Please check things in a positive way and do not exit methods in a more structured way. The best would be of course at the end of the method.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+			.promise()
+			.then(this.loadRoles);
+	};
+
+	/**
+	 * Removes given role to user.
+	 * @param {string} roleId Role identifier
+	 */
+	removeRoleFromUser = (roleId) => {
+		if (!confirm(this.context.i18n.getMessage('UserRoleEditor.Confirmation.delete'))) {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: Please check things in a positive way and do not exit methods in a more structured way. The best would be of course at the end of the method.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+			if (this.loadAssignableRolesPromise) {
+				this.loadAssignableRolesPromise.cancel();
+			}
+		}
+	}
+
+	/**
+	 * Handles role switching.
+	 */
+	onRoleChange = (option) => {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: If we're trying to use class methods please make them class methods and not closures.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	componentDidMount() {
+		this.loadRoles();
+	}
+
+	/**
+	 * Loads owned and assignable roles from api.
+	 */
+	loadRoles = () => {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: If we're trying to use class methods please make them class methods and not closures.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+		this.addRoleToUser(this.state.selectedRoleId)
+			.then(() => this.setState({ selectedRoleId: null }));
+	};
+
+	/**
+	 * Assigns given role to user.
+	 * @param {string} roleId Role identifier
+	 */
+	addRoleToUser = (roleId) => {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: If we're trying to use class methods please make them class methods and not closures.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+import UserRoleTable from './UserRoleTable.react';
+import UserRoleTableItem from './UserRoleTableItem.react';
+import Button from 'react-bootstrap/lib/Button';
+import Select from '@opuscapita/react-select';
+import './UserRoleEditor.css';
+
+class UserRoleEditor extends Component {
+
+	static propTypes = {
+		actionUrl: React.PropTypes.string.isRequired,
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: Please uses the new **prop-types** package and replace React.PropTypes with these PropTypes in order to not get deprecation warnings. (https://www.npmjs.com/package/prop-types)

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+				}
+			})
+			.then(() => this.setState({ isLoaded: true }));
+	};
+
+	/**
+	 * Add role form submit handler
+	 * @param {object} event Form submit event object
+	 */
+	onSubmit = (event) => {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: If we're trying to use class methods please make them class methods and not closures.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+			.set('Content-type', 'application/json')
+			.promise()
+			.then(this.loadRoles);
+	};
+
+	/**
+	 * Removes given role to user.
+	 * @param {string} roleId Role identifier
+	 */
+	removeRoleFromUser = (roleId) => {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: If we're trying to use class methods please make them class methods and not closures.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+	 */
+	onRoleChange = (option) => {
+		this.setState({ selectedRoleId: option.value });
+	};
+
+	/**
+	 * Determines whether add button should be disabled or not
+	 * @returns {boolean|*}
+	 */
+	get isAddRoleButtonDisabled() {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: I don't think we should use methods that are then used like variables. This is still somehow confusing in JS. Use the variables directly or call the method as method.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+		return !this.state.selectedRoleId;
+	}
+
+	/**
+	 * Roles which can be currently assigned.
+	 * @returns {string[]}
+	 */
+	get assignableRoles() {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: I don't think we should use methods that are then used like variables. This is still somehow confusing in JS. Use the variables directly or call the method as method.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js (hunk 8 above)
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: We're trying not to use **actionUrl**. In all service redesigns it is not available anymore. There is no such thing as a prefix anymore as the service to call is always available as */service-name*. So please remove **actionUrl** everywhere.

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+	/**
+	 * Assigns given role to user.
+	 * @param {string} roleId Role identifier
+	 */
+	addRoleToUser = (roleId) => {
+		return request
+			.put(`${this.props.actionUrl}/user/users/${encodeURIComponent(this.props.userId)}/roles/${roleId}`)
+			.set('Accept', 'application/json')
+			.set('Content-type', 'application/json')
+			.promise()
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: .promise() is not required if there is a call to .then().

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+	removeRoleFromUser = (roleId) => {
+		if (!confirm(this.context.i18n.getMessage('UserRoleEditor.Confirmation.delete'))) {
+			return;
+		}
+
+		return request
+			.delete(`${this.props.actionUrl}/user/users/${encodeURIComponent(this.props.userId)}/roles/${roleId}`)
+			.set('Accept', 'application/json')
+			.set('Content-type', 'application/json')
+			.promise()
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: .promise() is not required if there is a call to .then().

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+	/**
+	 * Loads owned and assignable roles from api.
+	 */
+	loadRoles = () => {
+		this.setState({ isLoaded: false });
+
+		this.loadOwnedRolesPromise = request
+			.get(`${this.props.actionUrl}/user/users/${encodeURIComponent(this.props.userId)}`)
+			.set('Accept', 'application/json')
+			.promise()
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: .promise() is not required if there is a call to .then().

File: src/client/components/UserRoleEditor/UserRoleEditor.react.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+			.then(response => this.setState({ownedRoles: response.body.roles}));
+
+		this.loadAssignableRolesPromise = request
+			.get(`${this.props.actionUrl}/user/users/current/assignableRoles`)
+			.set('Accept', 'application/json')
+			.promise()
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/UserRoleEditor.react.js]: .promise() is not required if there is a call to .then().

File: package.json [hunk 17]
```diff
@@ -16,29 +16,47 @@
     "exclude": []
   },
   "dependencies": {
+    "@opuscapita/i18n": "^1.1.0",
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply package.json]: Do you mean to move all frontend dependencies to dev or just @opuscapita/i18n?

File: local/index.html [hunk 18]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<title>User Editor</title>
+	<meta charset="utf-8">
+	<meta http-equiv="X-UA-Compatible" content="IE=edge">
+	<meta name="viewport" content="width=device-width, initial-scale=1">
+	<link rel="stylesheet" type="text/css" href="/user/static/css/jquery-ui-bundle.css">
+	<link rel="stylesheet" type="text/css" href="/user/static/css/ocui-bootstrap-bundle-0.1.2.css">
```
Comment (kwierchris) [inline local/index.html]: With service-base-ui you can remove this style.

File: local/index.html [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+	<link rel="stylesheet" type="text/css" href="/user/static/css/jcatalog.css">
```
Comment (kwierchris) [inline local/index.html]: This should be removed.

File: local/index.html [hunk 20]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<title>User Editor</title>
+	<meta charset="utf-8">
+	<meta http-equiv="X-UA-Compatible" content="IE=edge">
+	<meta name="viewport" content="width=device-width, initial-scale=1">
+	<link rel="stylesheet" type="text/css" href="/user/static/css/jquery-ui-bundle.css">
```
Comment (kwierchris) [inline local/index.html]: Are you using jquery UI?

File: src/client/components/UserRoleEditor/i18n/en.js [hunk 21]
```diff
@@ -0,0 +1,25 @@
+export default {
```
Comment (kwierchris) [inline src/client/components/UserRoleEditor/i18n/en.js]: See de.js

File: src/server/routes/index.js [hunk 22]
```diff
@@ -48,14 +48,19 @@ module.exports.init = function(app, db, config)
         app.put(['/api/users/current/profile', '/users/current/profile'], (req, res) => this.addOrUpdateUserProfile(req, res, true));
         app.put(['/api/users/:id/profile', '/users/:id/profile'], (req, res) => this.addOrUpdateUserProfile(req, res));
 
+        app.get('/users/current/assignableRoles', (req, res) => this.sendUserAssignableRoles(req, res, true));
```
Comment (kwierchris) [inline src/server/routes/index.js]: /api/users/...

File: src/server/routes/index.js [hunk 23]
```diff
@@ -48,14 +48,19 @@ module.exports.init = function(app, db, config)
...
+        app.get('/users/:id/assignableRoles', (req, res) => this.sendUserAssignableRoles(req, res));
```
Comment (kwierchris) [inline src/server/routes/index.js]: /api/users/...

File: src/server/routes/index.js [hunk 24]
```diff
@@ -48,14 +48,19 @@ module.exports.init = function(app, db, config)
...
+
         app.get(['/api/roles', '/roles'], (req, res) => this.sendRoles(req, res));
         app.get(['/api/roles/:id', '/roles/:id'], (req, res) => this.sendRole(req, res));
 
         app.post(['/api/roles', '/roles'], (req, res) => this.addUserRoles(req, res));
 
         app.put(['/api/users/:userId/roles/:roleId', '/users/:userId/roles/:roleId'], (req, res) => this.addUserToRole(req, res));
+
+        app.delete('/users/:userId/roles/:roleId', this.removeUserFromRole.bind(this))
```
Comment (kwierchris) [inline src/server/routes/index.js]: /api/users/...

File: src/server/static/css/jcatalog.css [hunk 25]
```diff
@@ -0,0 +1,27 @@
+footer .copyright {
```
Comment (kwierchris) [inline src/server/static/css/jcatalog.css]: Is this only for local testing?
Whatever, please don't use the name jcatalog anymore.

File: webpack.production.config.js [hunk 26]
```diff
@@ -0,0 +1,67 @@
+const path = require('path');
+const webpack = require('webpack');
+
+module.exports = {
+	entry: {
+		registration: './src/client/components/SupplierRegistrationEditor/index.js',
```
Comment (kwierchris) [inline webpack.production.config.js]: I guess, this should not be there.

**Pull request merged**